pgxn/neon/walsender_hooks.c

```diff
@@ -0,0 +1,121 @@
+#include "access/xlog.h"
+#include "access/xlog_internal.h"
+#include "access/xlogreader.h"
+#include "miscadmin.h"
+#include "utils/wait_event.h"
+
+#include "neon_walreader.h"
+#include "walproposer.h"
+
+static NeonWALReader *wal_reader = NULL;
+extern XLogRecPtr WalSndWaitForWal(XLogRecPtr loc);
+extern bool GetDonorShmem(XLogRecPtr *donor_lsn);
+
+static XLogRecPtr NeonWALReadWaitForWAL(XLogRecPtr loc)
+{
+    while(!NeonWALReaderUpdateDonor(wal_reader))
+        pg_usleep(1000);
+
+    return WalSndWaitForWal(loc);
+}
+
+static int NeonWALPageRead(
+    XLogReaderState *xlogreader,
+    XLogRecPtr targetPagePtr,
+    int reqLen,
+    XLogRecPtr targetRecPtr,
+    char *readBuf)
+{
+    // Wait for flush pointer to advance past our request
+    XLogRecPtr flushptr = NeonWALReadWaitForWAL(targetPagePtr + reqLen);
+    int count;
+    if (flushptr < targetPagePtr + reqLen)
+	return -1;
+
+    // Read at most XLOG_BLCKSZ bytes
+    if (targetPagePtr + XLOG_BLCKSZ <= flushptr)
+	count = XLOG_BLCKSZ;
+    else
+	count = flushptr - targetPagePtr;
+
+    // Sometimes walsender requests non-monotonic sequences of WAL. If that's
+    // the case, we have to reset streaming from remote at the correct position.
+    if(targetPagePtr != NeonWALReaderGetRemLsn(wal_reader))
+    {
+	NeonWALReaderResetRemote(wal_reader);
+    }
+
+    static WaitEventSet *wait_event_set;
+    for(;;)
+    {
+	NeonWALReadResult res = NeonWALRead(
+	    wal_reader,
+	    readBuf,
+	    targetPagePtr,
+	    count,
+	    walprop_pg_get_timeline_id());
+	if(res == NEON_WALREAD_SUCCESS)
+	{
+            // Free event set, since we'll have to recreate it for a new socket later anyway
+	    if (NeonWALReaderEvents(wal_reader) == 0 && wait_event_set != NULL)
+	    {
+		FreeWaitEventSet(wait_event_set);
+		wait_event_set = NULL;
+	    }
+            // Required by xlogreader
+	    xlogreader->seg.ws_tli = NeonWALReaderGetSegment(wal_reader)->ws_tli;
+	    xlogreader->seg.ws_segno = NeonWALReaderGetSegment(wal_reader)->ws_segno;
+	    xlogreader->seg.ws_file = NeonWALReaderGetSegment(wal_reader)->ws_file;
+	    return count;
+	}
+	if(res == NEON_WALREAD_ERROR)
+	{
+	    return -1;
+	}
+        // Res is WOULDBLOCK, so we wait on the socket, recreating event set if necessary
+	{
+	    pgsocket sock = NeonWALReaderSocket(wal_reader);
+	    WaitEvent event;
+	    long timeout_ms = 1000;
+
+	    if (wait_event_set == NULL)
+	    {
+		wait_event_set = CreateWaitEventSet(TopMemoryContext, 3);
+		AddWaitEventToSet(wait_event_set, WL_LATCH_SET, PGINVALID_SOCKET, MyLatch, NULL);
+		AddWaitEventToSet(wait_event_set, WL_EXIT_ON_PM_DEATH, PGINVALID_SOCKET, NULL, NULL);
+		AddWaitEventToSet(wait_event_set, WL_SOCKET_WRITEABLE | WL_SOCKET_READABLE, sock, NULL, NULL);
+	    }
+	    WaitEventSetWait(wait_event_set, timeout_ms, &event, 1, WAIT_EVENT_WAL_SENDER_MAIN);
+	}
+    }
+}
+
+static void NeonWALReadSegmentOpen(XLogReaderState *xlogreader, XLogSegNo nextSegNo, TimeLineID *tli_p)
```

Comment:
Yes I guess so, these callbacks seem to only be used by wal_read itself in the normal walsender. Should I remove it? I figure it's better to still fill out the callbacks struct.

Reply:
I'd remove it because it is confusing (we don't really open segment when reading remotely). Can leave callbacks which would panic if they are called, as an idea.

Reply:
Sure, made it panic.

Reply:
Ok actually panic doesn't work, the callback gets called by xlogreader in `WALRead`.